File: modules/coupon-api-service/src/main/java/com/thanh0x/coursedeal/exception/GlobalExceptionHandler.java
package com.thanh0x.coursedeal.exception;

import com.thanh0x.coursedeal.dto.ApiErrorDTO;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.context.request.ServletWebRequest;
import org.springframework.web.context.request.WebRequest;
import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;

import java.util.Date;

/**
 * GlobalExceptionHandler class handles exceptions globally within the application.
 */
@ControllerAdvice
public class GlobalExceptionHandler extends ResponseEntityExceptionHandler {

    @ExceptionHandler(BadRequestException.class)
    public ResponseEntity<ApiErrorDTO> handleBadRequestException(BadRequestException exception, WebRequest request) {
        return createErrorResponse(HttpStatus.BAD_REQUEST, exception.getMessage(), request);
    }

    @ExceptionHandler(ResourceNotFoundException.class)
    public ResponseEntity<ApiErrorDTO> handleResourceNotFoundException(ResourceNotFoundException exception, WebRequest request) {
        return createErrorResponse(HttpStatus.NOT_FOUND, exception.getMessage(), request);
    }

    @ExceptionHandler(UnsupportedOperationException.class)
    public ResponseEntity<ApiErrorDTO> handleUnsupportedOperationException(UnsupportedOperationException exception, WebRequest request) {
        return createErrorResponse(HttpStatus.NOT_IMPLEMENTED, exception.getMessage(), request);
    }

    @ExceptionHandler(Exception.class)
    public ResponseEntity<ApiErrorDTO> handleGlobalException(Exception exception, WebRequest request) {
        logger.error("Unexpected error: ", exception);
        return createErrorResponse(HttpStatus.INTERNAL_SERVER_ERROR, exception.getMessage(), request);
    }

    private ResponseEntity<ApiErrorDTO> createErrorResponse(HttpStatus status, String message, WebRequest request) {
        ApiErrorDTO error = ApiErrorDTO.builder()
                .timestamp(new Date())
                .status(status.value())
                .error(status.getReasonPhrase())
                .message(message)
                .path(((ServletWebRequest) request).getRequest().getRequestURI())
                .build();
        return new ResponseEntity<>(error, status);
    }
}

File: modules/coupon-api-service/src/main/kotlin/com/thanh0x/coursedeal/exception/GlobalExceptionHandler.kt
package com.thanh0x.coursedeal.exception

import com.thanh0x.coursedeal.config.logger
import com.thanh0x.coursedeal.dto.ApiErrorDTO
import jakarta.servlet.http.HttpServletRequest
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.context.request.ServletWebRequest
import org.springframework.web.context.request.WebRequest
import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler
import java.util.*

/**
 * GlobalExceptionHandler class handles exceptions globally within the application.
 */
@ControllerAdvice
class GlobalExceptionHandler : ResponseEntityExceptionHandler() {
    private val log = logger()

    @ExceptionHandler(BadRequestException::class)
    fun handleBadRequestException(exception: BadRequestException, request: WebRequest): ResponseEntity<ApiErrorDTO> {
        return createErrorResponse(HttpStatus.BAD_REQUEST, exception.message ?: "Bad Request", request)
    }

    @ExceptionHandler(ResourceNotFoundException::class)
    fun handleResourceNotFoundException(exception: ResourceNotFoundException, request: WebRequest): ResponseEntity<ApiErrorDTO> {
        return createErrorResponse(HttpStatus.NOT_FOUND, exception.message ?: "Not Found", request)
    }

    @ExceptionHandler(UnsupportedOperationException::class)
    fun handleUnsupportedOperationException(exception: UnsupportedOperationException, request: WebRequest): ResponseEntity<ApiErrorDTO> {
        return createErrorResponse(HttpStatus.NOT_IMPLEMENTED, exception.message ?: "Not Implemented", request)
    }

    @ExceptionHandler(Exception::class)
    fun handleGlobalException(exception: Exception, request: WebRequest): ResponseEntity<ApiErrorDTO> {
        log.error("Unexpected error: ", exception)
        return createErrorResponse(HttpStatus.INTERNAL_SERVER_ERROR, exception.message ?: "Internal Server Error", request)
    }

    private fun createErrorResponse(status: HttpStatus, message: String, request: WebRequest): ResponseEntity<ApiErrorDTO> {
        val error = ApiErrorDTO(
            timestamp = Date(),
            status = status.value(),
            error = status.reasonPhrase,
            message = message,
            path = (request as ServletWebRequest).request.requestURI
        )
        return ResponseEntity(error, status)
    }
}
